Change: Split ScriptDate:: into Calendar and Economy variants

2TallTyler opened the pull request:
## Motivation / Problem

#10700 splits off economy and calendar time. #11341 makes these time units not follow each other, and #11428 allows them to tick at different rates.

It's time we break the news to Game Scripts and AI.

## Description

This PR depends on and is built atop #10700.

This PR splits ScriptDate:: into ScriptDateCalendar:: and ScriptDateEconomy::, mirroring #10700's split of TimerGameCalendar and TimerGameEconomy.

All methods are duplicated, except for GetSystemTime(), which only exists in ScriptDateEconomy::. It could be split off into some new class if desired.

Internal calls to ScriptDate:: things are redirected to the correct class.

Okay, what about existing scripts? Well, the most crucial information for GS and AI is knowing how time in the game is progressing to grow towns, build new train lines, give the player new goals, etc. If calendar time can be frozen, those scripts break.

For this reason, we map all ScriptDate:: functions to ScriptDateEconomy::. More in this in Limitations.

## Limitations

- If a script looks for a specific calendar date rather than using dates only to measure the passing of time (I can't think of any examples, but a hypothetical example would be building industries based on real-world economic history) will behave strangely. They won't crash, but when OpenTTD is using real-time units, economy starts in year 1 and the script will react accordingly. I am not aware of any ways to split time without the potential of breaking scripts in this way. 😢 
- I have no idea how to test this or verify if my code is correct, since ScriptDate ignores the strong-type protection of our two time systems, and compat files are not built or verified by OpenTTD.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

2TallTyler committed:
Codechange: Sync economy and calendar dates when loading old saves

2TallTyler committed:
Codechange: Use ecomomy dates for timetables

2TallTyler committed:
Codechange: Use economy dates for industry production changes and closures

2TallTyler committed:
Codechange: Use economy dates for economy statistics and interest payments

2TallTyler committed:
Codechange: Use economy dates to clean up old news items

2TallTyler committed:
Codechange: Use economy dates for desync savegame interval

2TallTyler committed:
Codechange: Use economy dates for subsidy date displays

2TallTyler committed:
Codechange: Add economy date timer and variables

2TallTyler committed:
Codechange: Split daily/monthly/yearly timers into economy and calendar timers

2TallTyler committed:
Codechange: Use economy dates when drawing graphs

2TallTyler committed:
Codechange: Sync economy and calendar dates when starting a new game

2TallTyler committed:
Codechange: Sync economy and calendar dates when using date cheat

2TallTyler committed:
Codechange: Track the company age in economy years

2TallTyler committed:
Codechange: Use economy dates for vehicle service dates

2TallTyler committed:
Codechange: Use economy dates for linkgraphs

2TallTyler committed:
Doc: Clarify in economy and calendar timer files the usecases for each

2TallTyler committed:
Codechange: Move TimerGameCalendar functions to a parent TimerGameCommon class

2TallTyler committed:
Codechange: Update both calendar and economy dates when changing date in Scenario Editor

2TallTyler committed:
Codechange: Sync calendar and economy dates when loading NewGRFs

2TallTyler committed:
Codechange: Split ScriptDate file into Calendar and Economy files

2TallTyler committed:
Codechange: Split calls to ScriptDate into correct Calendar and Economy class

2TallTyler committed:
Fix: Add missing includes to timers from script implementation files

2TallTyler committed:
Change: Compatibility layers for AI and GS map ScriptDate:: to ScriptDateEcomomy::

rubidium42 commented on src/script/api/script_date_economy.hpp:
```diff
@@ -5,13 +5,13 @@
  */
 
-/** @file script_date.hpp Everything to query and manipulate date related information. */
+/** @file script_date_economy.hpp Everything to query and manipulate date related information. */
```
Given date stuff is split over two files, it's definitely not 'everything' anymore.

rubidium42 commented on src/script/api/script_date_economy.hpp:
```diff
@@ -5,13 +5,13 @@
+#include "../../timer/timer_game_economy.h"
 
 /**
  * Class that handles all date related (calculation) functions.
```
I would expect some documentation here (and in script_date_calendar.hpp), what is the difference between the two types of dates. Given you, as user, usually only see one I would expect this needs to be clarified for anyone writing an AI or GS.

rubidium42 reviewed (commented):
I'm not so sure whether mixing calendar and game time in the API in this way is useful for backward compatibility. You are, as seen from the API, returning vastly different dates that can't be reasonably compared.

Having said that... can't (old) scripts just always get the economy date and work just fine? Except when they code something to happen on a specific date of course.
Providing access to the calendar date would be useful for those specific instances, so maybe we can do a company/async mode-type trick? If the calendar-date-mode object is at the top, it returns and expects calendar dates, otherwise it will expect economy dates.

2TallTyler committed:
Codechange: LogCommandExecution writes economy date, not calendar date

Outside of Real-Time Mode, these are the same date, so debugging date-based events is still possible.

glx22 commented on src/script/api/script_date_economy.cpp:
```diff
@@ -0,0 +1,69 @@
+
+/* static */ ScriptDateEconomy::Date ScriptDateEconomy::GetCurrentDate()
+{
+	return (ScriptDateEconomy::Date)TimerGameCalendar::date.base();
```
I know you plan to do it differently, but looks like it's using the wrong timer, and it's the same in all other functions (except for parameter types).

2TallTyler replied:
Oops, faulty (user of) find-and-replace. But yes, I'm going to do this the other way.
